python/tank/bootstrap/async_bootstrap.py

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+# Import Qt without having to worry about the version to use.
+from ..util.qt_importer import QtImporter
+importer = QtImporter()
+if importer.wrapper is None:
+    # Raise an exception when Qt is not available.
+    raise ImportError
+QtCore = importer.QtCore
+QtGui = importer.QtGui
+
+from ..api import Sgtk
+
+
+class AsyncBootstrapWrapper(QtCore.QObject):
+    """
+    Wrapper class that can bootstrap an :class:`~sgtk.Sgtk` instance
+    asynchronously in a background thread, followed by the synchronous launching
+    of an :class:`~sgtk.platform.Engine` instance in the main application thread.
+    """
+
+    def __init__(self, toolkit_manager, engine_name, entity):
+        """
+        Initializes an instance of the asynchronous bootstrap wrapper.
+
+        This method retrieves the toolkit manager progress reporting callback to use it later.
+
+        :param toolkit_manager: :class:`~sgtk.bootstrap.ToolkitManager` instance bootstrapping the engine.
+        :param engine_name: Name of the engine to launch.
+        :param entity: Shotgun entity used to resolve a project context.
+        """
+
+        super(AsyncBootstrapWrapper, self).__init__()
+
+        self._toolkit_manager = toolkit_manager
+        self._engine_name = engine_name
+        self._entity = entity
+
+        # Retrieve the toolkit manager progress reporting callback.
+        self._progress_callback = self._toolkit_manager._progress_cb
+
+        self._completed_callback = None
+        self._failed_callback = None
```

Comment:
would also prevent from having to check to see if callback is defined before calling.